plugins/ccgx/fu-ccgx-dmc-device.c

```diff
@@ -0,0 +1,703 @@
+		return FALSE;
+	}
+	return TRUE;
+}
+
+
+static gboolean
+fu_ccgx_dmc_device_send_row_data (FuCcgxDmcDevice *self,
+				  guint8* row_buffer,
+				  guint16 row_size,
+				  GError **error)
+{
+	if (row_buffer == NULL) {
+		g_set_error(error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "invalid  pointer at %s",__func__);
+			return FALSE;
+	}
+	if (!g_usb_device_bulk_transfer (fu_usb_device_get_dev (FU_USB_DEVICE (self)),
+					 self->ep_bulk_out,
+					 row_buffer, row_size, NULL,
+					 DMC_BULK_OUT_PIPE_TIMEOUT,
+					 NULL, error)) {
+		g_prefix_error (error, "write row data error:");
+                return FALSE;
+	}
+        return TRUE;
+}
+
+
+static void
+fu_ccgx_dmc_device_to_string (FuDevice *device, guint idt, GString *str)
+{
+	FuCcgxDmcDevice *self = FU_CCGX_DMC_DEVICE (device);
+	fu_common_string_append_kv (str, idt, "UpdateModel",
+					fu_ccgx_dmc_update_model_to_string(self->update_model));
+	fu_common_string_append_kv (str, idt, "FwImageType",
+					fu_ccgx_fw_image_type_to_string (self->fw_image_type));
+	fu_common_string_append_kx (str, idt, "EpBulkOut", self->ep_bulk_out);
+	fu_common_string_append_kx (str, idt, "EpIntrIn", self->ep_intr_in);
+}
+
+static gboolean
+fu_ccgx_hpi_write_firmware_image (FuDevice *device,
+				  FuCcgxDmcFirmwareImageRecord *img_rcd,
+				  GError **error)
+{
+	FuCcgxDmcDevice *self = FU_CCGX_DMC_DEVICE (device);
+	GPtrArray *seg_records = NULL;
+	DmcIntRqt dmc_int_req = {0};
+
+	if (img_rcd == NULL) {
+		g_set_error(error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "invalid  pointer at %s",__func__);
+			return FALSE;
+	}
+	/* get segment records */
+	seg_records = img_rcd->seg_records;
+	for (guint32 seg_index = 0; seg_index < seg_records->len; seg_index++) {
+		GPtrArray *data_records = NULL;
+		FuCcgxDmcFirmwareSegmentRecord *seg_rcd = g_ptr_array_index (seg_records, seg_index);
+		/* write start row and number of rows to a device */
+		if (!fu_ccgx_dmc_device_send_write_command (self,seg_rcd->info_header.start_row,
+						seg_rcd->info_header.num_rows,
+						error))
+			return FALSE;
+		/* get data records */
+		data_records = seg_rcd->data_records;
+		for (guint32 data_index = 0; data_index < data_records->len; data_index++) {
+			guint8* row_buffer = NULL;
+			gsize row_size = 0;
+			/* get data from records */
+			FuCcgxDmcFirmwareDataRecord *data_rcd = g_ptr_array_index (data_records, data_index);
+			/* get row buffer from data */
+			row_buffer = g_bytes_get_data (data_rcd->data, &row_size);
+			/* write row data */
+			if (!fu_ccgx_dmc_device_send_row_data (self, row_buffer, (guint16)row_size, error))
+				return FALSE;
+			/* increase fw written size */
+			self->fw_data_written += row_size;
+			/* update update progress */
+			fu_device_set_progress_full (device, (gsize) self->fw_data_written, (gsize) self->fw_data_size);
+			/* get reponse  */
+			for (guint8 retry = 0; retry < 2; retry++) {
```

Comment:
I think you want fu_device_retry here.

Reply:
Apply fu_device_retry on today commit (ccgx: fix code alignment, modify code for review)